improve RGB support for tradfri

spektren opened the pull request:
Added feature detection (SUPPORT_RGB_COLOR) for IKEA tradfri RGB bulbs.

String comparison with manufacturer/device id used because I think there isn't a suitable property/param to identify the RGB capability. Both IKEA WS and RGB have data in the same parameters (x, y, rgb).

Added on the fly conversion of xyb values to rgb list to update method resulting in: 
    - bulb icon in UI is color scaled with current color (tested with IKEA WS+RGB and FLS-pp RGB) 
    - UI color picker or color temperature slider is displayed for respective bulb types

see also:
home-assistant#9603

Tested with the following setup:
Home Assistant 0.53.1
pytradfri 2.2
IKEA tradfri gateway fw 1.1.0015
zigbee rgb PWM module (dresden elektronik FLS-PP lp)
IKEA RGB, WS and dimmable-only bulbs

## Description:


**Related issue (if applicable):** fixes #9603 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#9411

## Example entry for `configuration.yaml` (if applicable):
```yaml
scene:
  - name: daytime
    entities:
      light.kitchen:
        state: on
        brightness: 63
        rgb_color: [255, 191, 0]
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ y ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ na ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ na ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ na ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ na ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ na ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

spektren committed:
Update tradfri.py

Added feature detection (SUPPORT_RGB_COLOR) for IKEA tradfri RGB bulbs.

String comparison with manufacturer/device id used because I think there isn't a suitable property/param to identify the RGB capability. Both IKEA WS and RGB have data in the same parameters (x, y, rgb).

Color picker in UI is only offered after one of the default RGB colors (i.e. efd275) is set in param 5706. For now I couldn't find the section in polymer where the decision (to show a color picker or not) is made.

spektren committed:
Update tradfri.py

spektren committed:
Update tradfri.py

spektren committed:
Update tradfri.py

pvizeli commented on homeassistant/components/light/tradfri.py:
```diff
@@ -107,10 +107,18 @@ def __init__(self, light, api):
+        _LOGGER.debug(
+            "features detected for %s %s: %d",
+            self._name, self._light.device_info.model_number, self._features
+        )
```
You see the feature on UI, there need no debug log...

spektren replied:
removed logging of detected features

MartinHjelmare commented:
The color selector will be shown when the RGB property of the entity does not return `None` after a state update.

spektren committed:
improve support of RGB bulbs

- debug logging of detected features removed (change requested)
- on the fly conversion of xyb values to rgb list added to update method
    resulting in: 
    - current rgb values will made be available for all bulbs/entities from current xyb values (i.e. IKEA WS and RGB) 
    - bulb icon in UI is color scaled with current color (tested with IKEA WS+RGB and FLS-pp RGB) 
    - UI color picker or color temperature slider is displayed for respective bulb types

spektren committed:
improve support of RGB bulbs with tradfri

- debug logging of detected features removed (change requested)
- on the fly conversion of xyb values to rgb list added to update method
    resulting in: 
    - current rgb values will made be available for all bulbs/entities from current xyb values (i.e. IKEA WS and RGB) 
    - bulb icon in UI is color scaled with current color (tested with IKEA WS+RGB and FLS-pp RGB) 
    - UI color picker or color temperature slider is displayed for respective bulb types

spektren committed:
Update tradfri.py

balloob commented:
I don't want to merge this before #7815 gets in, which is a major refactor of Tradfri and should land soon.

spektren commented:
Reincarnated in #9717
